### ISSUE TITLE
database or disk is full

### ISSUE THREAD

**Issue by dongmingh** [hyperledger-archives/fabric#718]
The database or disk is full when I ran an overnight stress test with invoke.

From the obcca-server window:
ERROR: 2016/02/23 11:53:31 ca.go:205: database or disk is full
ERROR: 2016/02/23 11:53:31 tca.go:252: database or disk is full
ERROR: 2016/02/23 11:53:31 ca.go:205: database or disk is full
ERROR: 2016/02/23 11:53:31 tca.go:252: database or disk is full

From the obc-peer window:
11:53:31.228 [crypto.client.lukas] NewChaincodeExecute -> ERRO 2fb626 Failed getting next transaction certificate [Failed loading TCerts from TCA].
11:53:31.228 [rest] Invoke -> ERRO 2fb627 {"Error": "Invoking Chaincode -- Failed loading TCerts from TCA"}
11:53:31.232 [crypto.client.lukas] callTCACreateCertificateSet -> ERRO 2fb628 Failed requesting tca create certificate set [rpc error: code = 2 desc = "database or disk is full"].
11:53:31.232 [crypto.client.lukas] NewChaincodeExecute -> ERRO 2fb629 Failed getting next transaction certificate [Failed loading TCerts from TCA].
11:53:31.232 [rest] Invoke -> ERRO 2fb62a {"Error": "Invoking Chaincode -- Failed loading TCerts from TCA"}
11:53:31.236 [crypto.client.lukas] callTCACreateCertificateSet -> ERRO 2fb62b Failed requesting tca create certificate set [rpc error: code = 2 desc = "database or disk is full"].
11:53:31.237 [crypto.client.lukas] NewChaincodeExecute -> ERRO 2fb62c Failed getting next transaction certificate [Failed loading TCerts from TCA].
11:53:31.237 [rest] Invoke -> ERRO 2fb62d {"Error": "Invoking Chaincode -- Failed loading TCerts from TCA"}
11:53:31.243 [crypto.client.lukas] callTCACreateCertificateSet -> ERRO 2fb62e Failed requesting tca create certificate set [rpc error: code = 2 desc = "database or disk is full"].
11:53:31.243 [crypto.client.lukas] NewChaincodeExecute -> ERRO 2fb62f Failed getting next transaction certificate [Failed loading TCerts from TCA].
11:53:31.244 [rest] Invoke -> ERRO 2fb630 {"Error": "Invoking Chaincode -- Failed loading TCerts from TCA"}

The disk space information is as follow when this occurs:
vagrant@vagrant-ubuntu-trusty-64:/var/openchain/production/db$ du -sh
4.4G    .
vagrant@vagrant-ubuntu-trusty-64:/var/openchain/production/db$ ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 31549
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 10000
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 31549
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited

vagrant@vagrant-ubuntu-trusty-64:/var/openchain/production/.obcca$ ls -la
total 7715864
drwxr-xr-x 2 vagrant vagrant       4096 Feb 23 11:53 .
drwxr-xr-x 6 vagrant vagrant       4096 Feb 22 22:34 ..
-rw-r--r-- 1 vagrant vagrant        587 Feb 22 22:34 eca.cert
-rw-r--r-- 1 vagrant vagrant       5120 Feb 22 22:34 eca.db
-rw-r--r-- 1 vagrant vagrant        233 Feb 22 22:34 eca.priv
-rw-r--r-- 1 vagrant vagrant        190 Feb 22 22:34 eca.pub
-rw-r--r-- 1 vagrant vagrant         44 Feb 22 22:34 obc.aes
-rw-r--r-- 1 vagrant vagrant        233 Feb 22 22:34 obc.ecies
-rw-r--r-- 1 vagrant vagrant        587 Feb 22 22:34 tca.cert
-rw-r--r-- 1 vagrant vagrant 7900954624 Feb 23 09:52 tca.db
-rw-r--r-- 1 vagrant vagrant         68 Feb 22 22:34 tca.hmac
-rw-r--r-- 1 vagrant vagrant        233 Feb 22 22:34 tca.priv
-rw-r--r-- 1 vagrant vagrant        190 Feb 22 22:34 tca.pub
-rw-r--r-- 1 vagrant vagrant        591 Feb 22 22:34 tlsca.cert
-rw-r--r-- 1 vagrant vagrant       5120 Feb 22 22:34 tlsca.db
-rw-r--r-- 1 vagrant vagrant        233 Feb 22 22:34 tlsca.priv
-rw-r--r-- 1 vagrant vagrant        190 Feb 22 22:34 tlsca.pub

vagrant@vagrant-ubuntu-trusty-64:/var/openchain/production/.obcca$ du -sh tca.db
7.4G    tca.db

**Comment by thkzrl** [comment]
How many TCerts have been generated?

**Comment by dongmingh** [comment]
It runs 130 invokes per second and last for at least 6+ hours.  Plus one hour query test at 235 per second before invoke test. So, it has more than 2.8  millions invoke + 0.8 million query.

**Comment by thkzrl** [comment]
Well, for a database with about 2.8 million certificates that's about the size to expect.

For batches of TCerts we can store only the information required to recreate the certificates on the fly whenever needed and thereby making the DB grow a bit slower; I'll evaluate the impacts of this and issue a matching PR if the trade-offs are acceptable. But in essence (and by design of OBC) it grows linear with the number of TCerts issued.

**Comment by thkzrl** [comment]
Membership services is now handled by ARC.

**Comment by ibmmark** [comment]
@diegomasini what is the latest here?

**Comment by diegomasini** [comment]
@ibmmark @dongmingh @binhn as a way to alleviate this situation we can stop storing the whole TCert in the database, ideally we don't need to store anything, but in order to implement revocation there is some information that needs to be stored. By doing so we are just buying time before this error happen again. I was talking with @jeronimoirazabal and we think the best solution is to decouple the code from the current database choice (sqlite3). There should be a mechanism to configure which datastore to use. In the meantime, if you agree we can stop storing TCerts in the TCA.

**Comment by thkzrl** [comment]
As mentioned before, an easy way to reduce storage requirements for TCerts would be to store just the generator information for TCert batches and re-generate them on demand. Since TCerts tend to be allocated in larger batches of, say, 100 this would significantly reduce the storage requirements. Some information has to be stored both for revocation and also auditing, I believe.

The code should be decoupled from the actual DB already; just passing a different DB driver name to sql.Open should work since, as to my knowledge, no DB-specific commands are used. We could make this a parameter in the yaml file if it helps.

**Comment by JonathanLevi** [comment]
Given enough time, any database is going to fill up if we keep adding to it. If the requirement is to store more than 4GB, then we can indeed switch to a better DB, though I don’t want to jump to the ‘quick and easy’ fix just yet.

Currently, 3 types of solutions sprang to mind:
1. A **different DB**
   Switching to a different (bigger/better/stronger/…) DB should not be difficult. I agree with @thkzr again. We can switch to a different DB supporting more data, etc. Having it configurable, allowing hot-swapping, or otherwise.
2. **Rate Limit**, **Revocation** and **Cleanups**
   As  `TCerts` are intended to be short-lived (especially when compared to the Enrollment Certs), do we really to store that many TCerts, and at such a rate? Shall we add a DB cleanup thread, remove `TCerts` after a certain time (e.g., especially those that have expired)? We can also put a limit on the rate of generating certs in the first place - and with the DB clean-up thread, we might be able to live with the (4GB) DB limit or so.
3. **Reproducibility** and "**storing Requests**" not "**Responses**"
   In the trade-off between storage and computation, we can store less and (re)compute more (e.g., the standard encryption/decryption approach). I also like the proposed notion of being able to regenerate `TCerts`, if we store all the information/data provided for the generation of the `TCerts` (replay-ability, in the positive sense, making sure that it's reproducible).

Means that in the trade-off between calculation/computation and storage, we can store the request and regenerate the results. So for example, having the persistent storage of the TCertBatch generation request, might do the trick, and store less.

**Comment by JonathanLevi** [comment]
But before choosing one/any of the above solutions for fixing this issue, an bigger (=or ‘silly’) question spring to mind: Why does the system currently store TCerts in a DB in the first place? What was the requirement/design that required this? Auditing/security/speed...?

The thinking is that if there isn't a real use-case, keeping the 100 TPS figure in mind, and given that we are dealing with (cryptographically) signed certs, that membersrvc portion could be easily re-written to live happily without persisting (storing/loading) these TCerts to a physical DB at all.

Suggestions?

**Comment by JonathanLevi** [comment]
But before choosing one/any of the above solutions for fixing this issue, a bigger (= maybe ‘silly’) question springs to mind: **Why does the system currently store TCerts in a DB** in the first place? What was the requirement/design guideline? Was stored for Auditing/Security/Speed (Caching)?

The thinking is (as @diegomasini suggested above): if there isn't a real use-case, keeping the 100 TPS figure in mind, and given that we are dealing with (cryptographically) signed certs, that membersrvc portion could be easily re-written to live happily without persisting (storing/loading) these TCerts to a physical DB at all.

Suggestions?

**Comment by JonathanLevi** [comment]
In the interim, just so that we can address this (started looking at this very recently) we can also do it in a flexible way, so that by default we do NOT store 'TCerts' in the/a DB at all.

Then if needed (again, will be good to get a compelling reason/use-case for storing these), we can configure the system so that it does support the full-fledged persistence-layer (for TCerts or other constructs, as the case may be).

**Comment by ibmmark** [comment]
Thanks @JonathanLevi @thkzrl @diegomasini definitely validated some thoughts I had, and great identification.

**Comment by christo4ferris** [comment]
@ibmmark @JonathanLevi @mastersingh24 IMO, we've only addressed a symptom. The root cause is actually not the bloated TCert storing, but that we don't have an effective strategy for dealing with a long running system vis storage of the db. We've only prolonged things a bit by addressing the symptom. Not saying that removing the TCerts was a bad idea. It was necessary. However, we need a more effective strategy for long term operations.
